external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java

```diff
@@ -0,0 +1,114 @@
+
+                for (Map.Entry<TopicPartition, OffsetManager> entry : offsetManagers.entrySet()) {
+                    TopicPartition topicPartition = entry.getKey();
+                    OffsetManager offsetManager = entry.getValue();
+
+                    long latestTimeOffset = endOffsets.get(topicPartition);
+                    long earliestTimeOffset = beginningOffsets.get(topicPartition);
+
+                    long latestEmittedOffset = offsetManager.getLatestEmittedOffset();
+                    long latestCompletedOffset = offsetManager.getCommittedOffset();
+                    long spoutLag = latestTimeOffset - latestCompletedOffset;
+
+                    String metricPath = topicPartition.topic()  + "/partition_" + topicPartition.partition();
+                    ret.put(metricPath + "/" + "spoutLag", spoutLag);
+                    ret.put(metricPath + "/" + "earliestTimeOffset", earliestTimeOffset);
+                    ret.put(metricPath + "/" + "latestTimeOffset", latestTimeOffset);
+                    ret.put(metricPath + "/" + "latestEmittedOffset", latestEmittedOffset);
+                    ret.put(metricPath + "/" + "latestCompletedOffset", latestCompletedOffset);
+
+                    TopicMetrics topicMetrics = topicMetricsMap.get(topicPartition.topic());
+                    if(topicMetrics == null) {
+                        topicMetrics = new TopicMetrics();
+                        topicMetricsMap.put(topicPartition.topic(), topicMetrics);
+                    }
+
+                    topicMetrics.totalSpoutLag += spoutLag;
+                    topicMetrics.totalEarliestTimeOffset += earliestTimeOffset;
+                    topicMetrics.totalLatestTimeOffset += latestTimeOffset;
+                    topicMetrics.totalLatestEmittedOffset += latestEmittedOffset;
+                    topicMetrics.totalLatestCompletedOffset += latestCompletedOffset;
+                }
+
+                for(Map.Entry<String, TopicMetrics> e : topicMetricsMap.entrySet()) {
+                    String topic = e.getKey();
+                    TopicMetrics topicMetrics = e.getValue();
+                    ret.put(topic + "/" + "totalSpoutLag", topicMetrics.totalSpoutLag);
+                    ret.put(topic + "/" + "totalEarliestTimeOffset", topicMetrics.totalEarliestTimeOffset);
+                    ret.put(topic + "/" + "totalLatestTimeOffset", topicMetrics.totalLatestTimeOffset);
+                    ret.put(topic + "/" + "totalLatestEmittedOffset", topicMetrics.totalLatestEmittedOffset);
+                    ret.put(topic + "/" + "totalLatestCompletedOffset", topicMetrics.totalLatestCompletedOffset);
+                }
+                LOG.debug("Metrics Tick: value : {}", ret);
+                return ret;
+            } else {
+                LOG.info("Metrics Tick: Not enough data to calculate spout lag.");
+            }
+        } catch (Exception t) {
```

Comment:
Sorry, I wasn't being clear with my earlier comment: Why are we catching exceptions here at all?